rpc/chain/generic/block.go

```diff
@@ -0,0 +1,254 @@
+)
+
+// GenericSignedBlock is the interface that represents the block of a particular chain.
+//
+// This interface is generic over types A, S, P, please check GenericExtrinsicSignature for more
+// information about these generic types.
+//
+//nolint:revive
+type GenericSignedBlock[A, S, P any] interface {
+	GetGenericBlock() GenericBlock[A, S, P]
+	GetJustification() []byte
+}
+
+// GenericBlock is the interface that holds information about the header and extrinsics of a block.
+//
+// This interface is generic over types A, S, P, please check GenericExtrinsicSignature for more
+// information about these generic types.
+//
+//nolint:revive
+type GenericBlock[A, S, P any] interface {
+	GetHeader() types.Header
+	GetExtrinsics() []GenericExtrinsic[A, S, P]
+}
+
+// GenericExtrinsic is the interface that holds the extrinsic information.
+//
+// This interface is generic over types A, S, P, please check GenericExtrinsicSignature for more
+// information about these generic types.
+//
+//nolint:revive
+type GenericExtrinsic[A, S, P any] interface {
+	GetVersion() byte
+	GetSignature() GenericExtrinsicSignature[A, S, P]
+	GetCall() types.Call
+}
+
+// GenericExtrinsicSignature is the interface that holds the extrinsic signature information.
+//
+// This interface is generic over the following types:
+//
+// A - Signer, the default implementation for this is the types.MultiAddress type
+// which can support a variable number of addresses.
+//
+// S - Signature, the default implementation for this is the types.MultiSignature type which can support
+// multiple signature curves.
+//
+// P - PaymentFields (ChargeAssetTx in substrate), the default implementation for this is the DefaultPaymentFields which
+// holds information about the tip sent for the extrinsic.
+//
+//nolint:revive
+type GenericExtrinsicSignature[A, S, P any] interface {
+	GetSigner() A
+	GetSignature() S
+	GetEra() types.ExtrinsicEra
+	GetNonce() types.UCompact
+	GetPaymentFields() P
+}
+
+// SignedBlock implements the GenericSignedBlock interface.
+type SignedBlock[A, S, P any] struct {
+	Block         *Block[A, S, P] `json:"block"`
+	Justification []byte          `json:"justification"`
+}
+
+func (s *SignedBlock[A, S, P]) GetGenericBlock() GenericBlock[A, S, P] {
+	return s.Block
+}
+
+func (s *SignedBlock[A, S, P]) GetJustification() []byte {
+	return s.Justification
+}
+
+// Block implements the GenericBlock interface.
+type Block[A, S, P any] struct {
+	Header     types.Header          `json:"header"`
+	Extrinsics []*Extrinsic[A, S, P] `json:"extrinsics"`
+}
+
+//nolint:revive
+func (b *Block[A, S, P]) GetHeader() types.Header {
+	return b.Header
+}
+
+//nolint:revive
+func (b *Block[A, S, P]) GetExtrinsics() []GenericExtrinsic[A, S, P] {
+	var res []GenericExtrinsic[A, S, P]
+
+	for _, ext := range b.Extrinsics {
+		res = append(res, ext)
+	}
+	return res
+}
+
+// Extrinsic implements the GenericExtrinsic interface.
+type Extrinsic[A, S, P any] struct {
```

Comment:
If this is the new Extrinsic struct, what is your vision for the other Extrinsic in the types dir? 
Would a user form & send extrinsics using this type instead of the one the `types` dir? Would be great if we could unify them.

More over, is there any further clean up (we can do later anyways) that we can do on those types?

Reply:
This is a type used only for state reads, for now. I think that in the future we can use something like this when performing calls as well. If we decide on the amount of flexibility coming from the generics, we can also replace the current logic with these types and provide default builders for them.

Reply:
As agreed, let's tackle this topic in a follow-up to the grant work, after the grant is reviewed.